File: dsa-problems/Probabilistic/solutions/javascript/PRB-003-reservoir-sampling-k.js
const readline = require("readline");

function reservoirSample(n, k, seed) {
  if (k === 0) return [];
  if (k > n) k = n;

  const reservoir = new Int32Array(k);
  for (let i = 0; i < k; i++) {
    reservoir[i] = i + 1;
  }

  let state = BigInt(seed);
  const multiplier = 6364136223846793005n;
  const increment = 1n;
  const mask = (1n << 64n) - 1n;

  for (let i = k + 1; i <= n; i++) {
    state = (state * multiplier + increment) & mask;

    // BigInt modulo
    const j = state % BigInt(i);

    if (j < BigInt(k)) {
      reservoir[Number(j)] = i;
    }
  }

  return Array.from(reservoir);
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  const n = parseInt(data[0], 10);
  const k = parseInt(data[1], 10);
  const seed = BigInt(data[2]);
  const res = reservoirSample(n, k, seed);
  console.log(res.join(" "));
});

File: dsa-problems/Probabilistic/solutions/python/PRB-003-reservoir-sampling-k.py
import sys
import random

def reservoir_sample(n: int, k: int, seed: int):
    if k == 0:
        return []
    if k > n:
        k = n

    random.seed(seed)
    reservoir = list(range(1, k + 1))

    for i in range(k, n):
        j = random.randint(0, i)
        if j < k:
            reservoir[j] = i + 1

    return reservoir

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    n = int(data[0])
    k = int(data[1])
    seed = int(data[2])
    res = reservoir_sample(n, k, seed)
    res.sort()
    print(" ".join(str(x) for x in res))

if __name__ == "__main__":
    main()
